### PR TITLE
Refactor + Rebalance of species armor

### PR DESCRIPTION
**Pull request by XDTM**
:cl: XDTM
balance: Species armor has been rebalanced:
balance: Abductors are slightly more resistant to energy weapons and biological weapons.
balance: Androids are now radiation immune, and are slightly more resistant to melee, bullet and bomb damage, but are more affected by energy weapons.
balance: Flypeople take slightly more bullet and melee damage, but are more resistant to biological weapons.
balance: Golems no longer have a 55% reduction to all damage. They're now slightly more resistant to bullet, melee and bomb damage but much less resistant (but still resistant) to laser damage. Diamond, gold and wood golems have the same pattern, with higher, and lower values respectively.
balance: Sand and Glass golems are very weak against bomb damage.
balance: Slime and jelly people are now resistant to lasers, energy, and bomb damage, but are much easier to dismember. (No bones, after all)
balance: Skeletons now take increased damage from lasers and much increased damage from melee, bullets and bombs. However, there is a chance that bullets may pass through them harmlessly.
balance: Skeletons have instead increased armor if they have milk in their system.
/:cl:

All values are not final, i'll undo any controversial balance changes.

### CODE REVIEW

File: code/modules/mob/living/carbon/human/species.dm [hunk 1]
```diff
@@ -38,7 +38,7 @@
 	var/list/mutant_bodyparts = list() 	// Parts of the body that are diferent enough from the standard human model that they cause clipping with some equipment
 	var/list/mutant_organs = list()		//Internal organs that are unique to this race.
 	var/speedmod = 0	// this affects the race's speed. positive numbers make it move slower, negative numbers make it move faster
-	var/armor = 0		// overall defense for the race... or less defense, if it's negative.
+	var/list/armor = list(melee = 0, bullet = 0, laser = 0, energy = 0, bomb = 0, bio = 0, rad = 0, fire = 0, acid = 0) //species innate armor
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/human/species.dm]: don't define these like this, they should be "melee", "bullet", "laser" etc, the other way is non-standard and has caused us issues in the past.

File: code/modules/mob/living/carbon/human/species_types/abductors.dm [hunk 2]
```diff
@@ -4,6 +4,7 @@
 	say_mod = "gibbers"
 	sexes = 0
 	species_traits = list(NOBLOOD,NOBREATH,VIRUSIMMUNE,NOGUNS,NOHUNGER)
+	armor = list(melee = 0, bullet = 0, laser = 0, energy = 20, bomb = 0, bio = 50, rad = 50, fire = 0, acid = 0) //bizarre alien biology
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/human/species_types/abductors.dm]: ditto for the rest

**Comment by Armhulen** [comment]
>>>>>>no shadowpeople buff

Fuck right off

**Comment by QualityVan** [comment]
If you're gonna increase the damage skeletons take they need to be able to drink milk to improve their resistances.

**Comment by GunHog** [comment]
I feel you need to do some research on how bomb and bio "armor" works. Bio "armor" is nigh worthless and I still believe that permeability_coefficient controls some things that bio should.

Negative values for explosions do nothing. Also, bomb "armor" is chance based. See here how it works: https://github.com/tgstation/tgstation/blob/master/code/modules/mob/living/carbon/human/human_defense.dm#L361
Basically, bomb armor is the probability of taking half damage from light and heavy explosions, and being killed + thrown instead of gibbed in a devastating explosion.

File: code/modules/mob/living/carbon/human/species.dm (hunk 1 above)
Comment (XDTM) [inline code/modules/mob/living/carbon/human/species.dm]: All armor is currently defined like this, won't this require a large scale refactor?

**Comment by XDTM** [comment]
i know about gibchance, didn't know it was also chancebased for lower tiers. As for bio armor, yeah, i know it's useless but maybe i or someone will make it count and they'll have it then. Probably amount of chems absorbed from sprays and such, virus chance, special bioweapons like the decloner.

**Comment by XDTM** [comment]
Skeletons now gain extra armor when they have milk in their body.

**Comment by Bawhoppen** [comment]
but why

**Comment by XDTM** [comment]
It irked me that species had a single armor value to adjust that affected every type of damage at once, wanted to standardize it, got carried away

File: code/modules/mob/living/carbon/human/species.dm (hunk 1 above)
Comment (Cyberboss) [inline code/modules/mob/living/carbon/human/species.dm]: It's the same thing, but not having them as strings causes maintainability issues

**Comment by XDTM** [comment]
@RemieRichards i made a delete_chemicals proc for species

**Comment by Armhulen** [comment]
>Abductors are slightly more resistant to energy weapons and biological weapons.
buff they did not need but this is probably ok
>Androids are now radiation immune, and are slightly more resistant to melee, bullet and bomb damage, but are more affected by energy weapons.
most of the time this will be a huge boon, on one of the strongest races in the game right now
>Flypeople take slightly more bullet and melee damage, but are more resistant to biological weapons.
roundstart race, should stay vanilla like lizards
>Golems no longer have a 55% reduction to all damage. They're now slightly more resistant to bullet, melee and bomb damage but much less resistant (but still resistant) to laser damage. Diamond, gold and wood golems have the same pattern, with higher, and lower values respectively.
>>>20 MINUTES OF XENOBIOLOGY FOR GOLEMS THAT WILL DIE TO LIKE 3 LASERS
no
>Sand and Glass golems are very weak against bomb damage.
yes for sand, does glass REALLY NEED ANOTHER NERF.
>Slime and jelly people are now resistant to lasers, energy, and bomb damage, but are much easier to dismember. (No bones, after all)
actually pretty good
>Skeletons now take increased damage from lasers and much increased damage from melee, bullets and bombs. However, there is a chance that bullets may pass through them harmlessly.
>>>20 MINUTES OF XENOBIOLOGY FOR A RACE THAT KILLS YOU IF YOU DO ANYTHING
>Skeletons have instead increased armor if they have milk in their system.
i love greytiding skeletons, at least they're a joke to kill
>Shadowpeople get no changes
wtf

**Comment by Armhulen** [comment]
tl;dr please do not do this

**Comment by XDTM** [comment]
Golems have 35% laser resistance, which is still a lot.
Golem balancing also takes into account free golems, and the bomb stuff was more because glass shatters and sand scatters. As gunhog said, it might be pointless.
Skeletons were pretty much the go-to powergame race due to its immunities. Now it has a significant downside.

As for shadowpeople, i'm not sure if their benefits should lie in armor, i was thinking more esoteric shadowbased powers would be better than just being harder to kill.

File: code/modules/mob/living/carbon/human/species_types/golems.dm [hunk 3]
```diff
@@ -270,7 +270,7 @@
 	id = "sand golem"
 	fixed_mut_color = "ffdc8f"
 	meat = /obj/item/weapon/ore/glass //this is sand
-	armor = 0
+	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = -100, "bio" = 100, "rad" = 100, "fire" = 0, "acid" = 0)
 	burnmod = 3 //melts easily
 	brutemod = 0.25
 	info_text = "As a <span class='danger'>Sand Golem</span>, you are immune to physical bullets and take very little brute damage, but are extremely vulnerable to burn damage. You will also turn to sand when dying, preventing any form of recovery."
```
Comment (Armhulen) [inline code/modules/mob/living/carbon/human/species_types/golems.dm]: As a sand golem, you are immune to physical bullets and we're lying to you by saying you take very little brute when in reality you take full brute damage, fuck you

Comment (XDTM) [inline code/modules/mob/living/carbon/human/species_types/golems.dm]: read the brutemod

Comment (Armhulen) [inline code/modules/mob/living/carbon/human/species_types/golems.dm]: i assume that means heavy armor, okay okay

**Comment by Armhulen** [comment]
go for shadow powers sure, they need a lot of buffs considering
*pod people heal more in light than shadowpeople do in dark
*shadowpeople burn in dark while pod people will get hungry after awhile of it
*pod people never get hungry or full in the light, shadowpeople can starve in the darkness

**Comment by Armhulen** [comment]
And sure, while skeleton was extremely strong it isn't fair to do this considering you buffed android to be really strong now, having natural armor and only being weaker to energy weapons along with all its other upsides

**Comment by Armhulen** [comment]
And on the golems, I don't think they should get a nerf. Thinking about the balance in the ghost role sense is not a very good idea as this essentially nerfs the stationwide golems a lot and buffs the ghost role golems, the ones who most of the time will never see a human, a lot more.

**Comment by XDTM** [comment]
Alright, i'll adjust the values

**Comment by Armhulen** [comment]
i'm very adamant on races as i've been pushing for their balance for a very long time.

**Comment by Armhulen** [comment]
skeletons are still really weak after this change, my say is to make them -10 for all those debuffs, remove the weird 25% chance dodge and that'll do. My reasoning is that they're a big part of wizard balance, leaving -30 is a pretty big nerf to liches and thralls.

**Comment by Armhulen** [comment]
That being said, those changes were a step in the right direction.

**Comment by XDTM** [comment]
For two spell points, they get immunity to most environimental effects, immunity from chem syringes which are the second best counter aside from shotguns, and on top of that they get to revive from death, with a recent buff since it also works when the body is destroyed. Being more fragile is justified imo, and a competent lich can find milk for tough times.

**Comment by Armhulen** [comment]
Fair, 👍

**Comment by Lumipharon** [comment]
As above, skeletons don't need to be, nor should they be, tough. Also fuck dem liches.

**Comment by Armhulen** [comment]
they are one of the most expensive transformations, Lumi.

**Comment by Lumipharon** [comment]
Because it's an extremely powerful buff.

**Comment by Armhulen** [comment]
>extremely hard to get
>they should not be strong
thinkey man

**Comment by Lumipharon** [comment]
I'm more talking about lich, not xenomemes. Unless lich is not skeleton species?

**Comment by Armhulen** [comment]
liches are skeleton species.

**Comment by Armhulen** [comment]
but nerfing liches are nerfing xenomemes.

**Comment by XDTM** [comment]
I forgot to make zombies tougher, i'll add that when i can

**Comment by Fox-McCloud** [comment]
I support the system changes to this.

But I don't think you should be doing major armor balances changes in the same PR.

**Comment by Cyberboss** [comment]
What Fox said

**Comment by AnturK** [comment]
Yeah, please separate so bikeshedding can go on while the system gets in.

**Comment by XDTM** [comment]
oh god i have to figure out again how to undo commits now

**Comment by Armhulen** [comment]
I agree I think races are imbalanced right now but we should be balancing them with fun stuff like easy dismemberment not weirdo armor values

**Comment by lzimann** [comment]
Are you still doing this?

**Comment by Armhulen** [comment]
Still didn't atomize balance part with the refactor, shame!

**Comment by lzimann** [comment]
Just ping whenever you keep doing it(or I guess a new pr without balance changes, whatever you choose is fine).

**Pull request closed without merging**